### PR TITLE
Drop custom `__eq__` from `Status`

### PR DESCRIPTION
**Pull request by jakirkham**
This appears to eat up a good chunk of time in places where the `Status` needs to be checked. So try dropping this method in favor of the default implementation.

That said, I'm sure this is here for some reason. So marking this as WIP so we can discuss whether removing is ok or if we need to figure out some alternative solution.

### CODE REVIEW

**Comment by mrocklin** [comment]
cc @Carreau

**Comment by jakirkham** [comment]
For context this `Enum` was added in PR ( https://github.com/dask/distributed/pull/3853 ). AIUI this custom `__eq__` was about having a smooth change from using `str`s to `Enum`s in downstream libraries. Not sure if that already happened.

cc @lesteve @jacobtomlinson @jcrist

**Comment by jacobtomlinson** [comment]
From a search through the Dask org it looks like strings are still used in dask-cloudprovider and dask-jobqueue.
We will need to update those before merging this.

**Comment by jakirkham** [comment]
What sort of places are these `str`s used? Unfortunately searching for things like `"finished"` on GitHub were less helpful than expected. Though do see some use of [`Status` here and below]( https://github.com/dask/dask-cloudprovider/blob/09c28979fd283d483f46ffec7e5165d887a9083a/dask_cloudprovider/aws/tests/test_ec2.py#L91 ).

File: distributed/core.py [hunk 1]
```diff
@@ -61,32 +61,6 @@ class Status(Enum):
...
     dont_reply = "dont-reply"
 
-    def __eq__(self, other):
-        """
-        Implement equality checking with backward compatibility.
-
-        If other object instance is string, we compare with the values, but we
-        actually want to make sure the value compared with is in the list of
-        possible Status, this avoid comparison with non-existing status.
-        """
```
Comment (mrocklin) [inline distributed/core.py:71]: Given the need for some backwards compatibility, I wonder if we might first try the following:

```python
try:
    return self.value == other.value
except AttributeError:
    ...
```

My experience is that try-except tends to be quite a bit faster than if/else in the try case, and this avoids the type check as well (which should be fine).

We can also migrate over all of the other projects if that will provide a large boost.  It isn't hard to do.

**Comment by lesteve** [comment]
This should be easily fixable in dask-jobqueue, if I believe my original comment in https://github.com/dask/distributed/pull/3853#issuecomment-639282877. I have not done anything about it (and probably not planning to do anything about it in the next few days ...) though. Don't hesitate ping me if you open a PR in dask-jobqueue!

**Comment by mrocklin** [comment]
See https://github.com/dask/dask-jobqueue/pull/476

**Comment by mrocklin** [comment]
And https://github.com/dask/dask-cloudprovider/pull/200

Also https://github.com/dask/distributed/pull/4280 for the future

**Comment by jakirkham** [comment]
Thanks for filing all of these Matt! 😄

**Comment by jakirkham** [comment]
Looks like those PRs have been merged. Thanks all! 😄

Marking this as ready for review 🙂

**Comment by jakirkham** [comment]
Is anything else needed here? Or are we good to merge?

**Comment by mrocklin** [comment]
@jacobtomlinson @lesteve @guillaumeeb it would be convenient for the Scheduler performance work to drop backwards compatibility for string-valued statuses in the core dask.distributed package.  This would _break_ interoperability with dask-cloudprovider and dask-jobqueue prior to the recent PRs sent in last week.

Are you comfortable releasing soonish and dealing with the inevitable user complaints?

**Comment by jakirkham** [comment]
It's worth noting the `Enum` object has been around since `distributed` version `2.19.0` ( https://github.com/dask/distributed/commit/1408ab59e691378f5a5c0299c8ac0bb4f42d4b22 ).

I don't think we are planning on releasing Distributed soon, but I could be wrong about that.

**Comment by mrocklin** [comment]
Yeah, the enum has been around, but we didn't do a good job originally in
actually making the change throughout the subprojects consistently until
now.  This puts the users and maintainers of the subprojects in a
suboptimal position.

**Comment by jakirkham** [comment]
Yep I understand. Just sharing info for the sake of transparency.

**Comment by Carreau** [comment]
Sorry for the late reply; this does looks good to me; 

If you are a bit worries, you can make one release where you turn this warning on by default. It will raise on users' code but they will still be able to turn the error off with a warning filter to ignore.

**Comment by jacobtomlinson** [comment]
Dask cloudprovider 0.5.1 is now on PyPI. Thanks @mrocklin for pushing this forwards.

**Comment by jakirkham** [comment]
Dask-Jobqueue 0.7.2 has also been released.

Since it sounds like we are planning to do a release of Dask + Distributed soon. Maybe we can merge this after that release? That would allow for more uptake of the downstream projects' releases before this change shows up. Thoughts?

**Comment by mrocklin** [comment]
That seems like a good idea.  Thank you for being flexible with the timing
on this one.  I know you'd like to have it out of your profiles

**Comment by jakirkham** [comment]
Now that Dask + Distributed have been released and a week has passed, am curious are we comfortable merging this change? Or would we like to wait longer?

**Comment by jakirkham** [comment]
Thoughts @mrocklin? 🙂

**Comment by mrocklin** [comment]
Sure.  Let's go for it.

**Pull request merged**